src/librustc_mir/interpret/operand.rs

```diff
@@ -577,91 +577,112 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
     pub fn read_discriminant(
         &self,
         rval: OpTy<'tcx, M::PointerTag>,
-    ) -> InterpResult<'tcx, (u128, VariantIdx)> {
+    ) -> InterpResult<'tcx, (Scalar<M::PointerTag>, VariantIdx)> {
         trace!("read_discriminant_value {:#?}", rval.layout);
 
-        let (discr_layout, discr_kind, discr_index) = match rval.layout.variants {
+        let (discr_scalar_layout, discr_kind, discr_index) = match rval.layout.variants {
             Variants::Single { index } => {
-                let discr_val = rval
-                    .layout
-                    .ty
-                    .discriminant_for_variant(*self.tcx, index)
-                    .map_or(u128::from(index.as_u32()), |discr| discr.val);
-                return Ok((discr_val, index));
+                let discr = match rval.layout.ty.discriminant_for_variant(*self.tcx, index) {
+                    Some(discr) => {
+                        // This type actually has discriminants.
+                        let discr_layout = self.layout_of(discr.ty)?;
+                        Scalar::from_uint(discr.val, discr_layout.size)
+                    }
+                    None => {
+                        // On a type without actual discriminants, return variant idx as `u8`.
+                        let discr_layout = self.layout_of(self.tcx.types.u8)?;
+                        Scalar::from_uint(index.as_u32(), discr_layout.size)
+                    }
+                };
+                return Ok((discr, index));
             }
-            Variants::Multiple { discr: ref discr_layout, ref discr_kind, discr_index, .. } => {
-                (discr_layout, discr_kind, discr_index)
+            Variants::Multiple { ref discr, ref discr_kind, discr_index, .. } => {
+                (discr, discr_kind, discr_index)
             }
         };
 
-        // read raw discriminant value
-        let discr_op = self.operand_field(rval, discr_index)?;
-        let discr_val = self.read_immediate(discr_op)?;
-        let raw_discr = discr_val.to_scalar_or_undef();
-        trace!("discr value: {:?}", raw_discr);
-        // post-process
+        // There are *three* types/layouts that come into play here:
+        // - The field storing the discriminant has a layout, which my be a pointer.
+        //   This is `discr_val.layout`; we just use it for sanity checks.
```

Comment:
"storing the discriminant" could be something more like "*encoding* the discriminant as a tag/niche".

That might make it clearer why the type differs: the only reason to have a simple encoding is cheap decoding, it could be any arbitrary mapping as long as you have enough values for every possible discriminant to be represented.

Reply:
> "encoding the discriminant as a tag/niche".

But the layout of the tag/niche is `discr_layout`, right? The second item in my list? And then `discr_ty` is the high-level thing?

Reply:
> But the layout of the tag/niche is discr_layout, right? 

If this is correct I'll rename the variables to `tag_layout` or so.

Reply:
Did that, and also opened https://github.com/rust-lang/rust/pull/72497 for the general terminology cleanup.